File: packages/server/app/controllers/puzzle-controller/puzzle-controller.ts
import { WordPlacement } from '@common/models/word-finding';
import { BaseController } from '@app/controllers/base-controller';
import { wordPlacementValidator } from '@app/middlewares/validators/word-placement';
import { PuzzleService } from '@app/services/puzzle-service/puzzle.service';
import { UserRequest } from '@app/types/user';
import { Router } from 'express';
import { StatusCodes } from 'http-status-codes';
import { Service } from 'typedi';
import { Position } from '@app/classes/board';
import { PuzzleResultStatus } from '@common/models/puzzle';

@Service()
export class PuzzleController extends BaseController {
    constructor(private readonly puzzleService: PuzzleService) {
        super('/api/puzzles');
    }

    protected configure(router: Router): void {
        router.post('/start', async (req: UserRequest, res, next) => {
            try {
                res.status(StatusCodes.OK).send(await this.puzzleService.startPuzzle(req.body.idUser));
            } catch (e) {
                next(e);
            }
        });

        router.post('/daily/start', async (req: UserRequest, res, next) => {
            try {
                res.status(StatusCodes.OK).send(await this.puzzleService.startDailyPuzzle(req.body.idUser));
            } catch (e) {
                next(e);
            }
        });

        router.post('/daily/is-completed', async (req: UserRequest, res, next) => {
            try {
                res.status(StatusCodes.OK).send({ isCompleted: !(await this.puzzleService.canDoDailyPuzzle(req.body.idUser)) });
            } catch (e) {
                next(e);
            }
        });

        router.post('/daily/leaderboard', async (req: UserRequest, res, next) => {
            try {
                res.status(StatusCodes.OK).send(await this.puzzleService.getDailyPuzzleLeaderboard(req.body.idUser));
            } catch (e) {
                next(e);
            }
        });

        router.post(
            '/complete',
            ...wordPlacementValidator('wordPlacement'),
            async (req: UserRequest<{ wordPlacement: WordPlacement }>, res, next) => {
                try {
                    res.status(StatusCodes.OK).send(
                        await this.puzzleService.completePuzzle(req.body.idUser, {
                            ...req.body.wordPlacement,
                            startPosition: Position.fromJson(req.body.wordPlacement.startPosition),
                        }),
                    );
                } catch (e) {
                    next(e);
                }
            },
        );

        router.post('/abandon', async (req: UserRequest<{ status?: PuzzleResultStatus }>, res, next) => {
            try {
                res.status(StatusCodes.OK).send(await this.puzzleService.abandonPuzzle(req.body.idUser, req.body.status));
            } catch (e) {
                next(e);
            }
        });
    }
}

File: packages/mobile/lib/controllers/puzzle-controller.dart
import 'dart:convert';

import 'package:http/http.dart';
import 'package:http_interceptor/http/intercepted_http.dart';
import 'package:mobile/classes/actions/word-placement.dart';
import 'package:mobile/classes/puzzle/puzzle-result.dart';
import 'package:mobile/constants/endpoint.constants.dart';
import '../locator.dart';
import '../services/client.dart';

class PuzzleController {
  PersonnalHttpClient httpClient = getIt.get<PersonnalHttpClient>();

  InterceptedHttp get http => httpClient.http;

  final String endpoint = PUZZLE_ENDPOINT;

  PuzzleController._privateConstructor();

  Future<Response> startPuzzle() async {
    return http.post(Uri.parse("$endpoint/start"));
  }

  Future<Response> completePuzzle(WordPlacement placement) async {
    return http.post(Uri.parse("$endpoint/complete"), body: jsonEncode(placement));
  }

  Future<Response> abandonPuzzle({required PuzzleResultStatus resultStatus}) async {
    return http.post(Uri.parse("$endpoint/abandon"), body: jsonEncode(resultStatus));
  }

  void quitPuzzle() {
    http.post(Uri.parse("$endpoint/abandon"));
  }

  Future<Response> startDailyPuzzle() async {
    return http.post(Uri.parse("$endpoint/daily/start"));
  }

  Future<Response> isDailyCompleted() async {
    return http.post(Uri.parse("$endpoint/daily/is-completed"));
  }

  Future<Response> getDailyPuzzleLeaderboard() {
    return http.post(Uri.parse("$endpoint/daily/leaderboard"));
  }

  static final PuzzleController _instance =
      PuzzleController._privateConstructor();

  factory PuzzleController() {
    return _instance;
  }
}
